src/preprocessing/passes/bv_ackermann.cpp

```diff
@@ -41,11 +42,54 @@ namespace passes {
 namespace
 {
 
-void storeFunction(
-    TNode func,
-    TNode term,
-    FunctionToArgsMap& fun_to_args,
-    SubstitutionMap& fun_to_skolem)
+void addLemmaForPair(TNode args1,
+                     TNode args2,
+                     const TNode func,
+                     AssertionPipeline* assertionsToPreprocess,
+                     NodeManager* nm,
+                     std::stack<TNode>* stack)
+{
+  Node args_eq;
+
+  if (args1.getKind() == kind::APPLY_UF)
+  {
+    AlwaysAssert(args1.getKind() == kind::APPLY_UF
+                 && args1.getOperator() == func);
+    AlwaysAssert(args2.getKind() == kind::APPLY_UF
+                 && args2.getOperator() == func);
+    AlwaysAssert(args1.getNumChildren() == args2.getNumChildren());
+
+    std::vector<Node> eqs(args1.getNumChildren());
+
+    for (unsigned i = 0, n = args1.getNumChildren(); i < n; ++i)
+    {
+      eqs[i] = nm->mkNode(kind::EQUAL, args1[i], args2[i]);
+    }
+    args_eq = bv::utils::mkAnd(eqs);
+  }
+  else
+  {
+    AlwaysAssert(args1.getKind() == kind::SELECT && args1[0] == func);
+    AlwaysAssert(args2.getKind() == kind::SELECT && args2[0] == func);
+    AlwaysAssert(args1.getNumChildren() == 2);
+    AlwaysAssert(args2.getNumChildren() == 2);
+    args_eq = nm->mkNode(kind::EQUAL, args1[1], args2[1]);
+  }
+  Node func_eq = nm->mkNode(kind::EQUAL, args1, args2);
+  Node lemma = nm->mkNode(kind::IMPLIES, args_eq, func_eq);
+  assertionsToPreprocess->push_back(lemma);
+  /* add the constraint to the stack so that
+  * collectFunctionsAndLemmas will process it as well. */
+  stack->push(lemma);
+}
+
+void storeFunctionAndAddLemmas(TNode func,
+                               TNode term,
+                               FunctionToArgsMap& fun_to_args,
+                               SubstitutionMap& fun_to_skolem,
+                               AssertionPipeline* assertions,
+                               NodeManager* nm,
```

Comment:
Since you're passing in the node manager, why not use it below (instead of calling NodeManager::currentNM())?